Updating Stub Format for VMSMERouter Changes

mlarson02 opened the pull request:
This PR updates the DTC emulation and Track Finding emulation Disk Stub data format by adding to the beginning of the stubword a bit determining whether the stub is in the negative z region of the detector, and adding an offset to the encoded r value for DiskPS stubs of 256 (7.5 cm) to allow fitting the negDisk bit in 36 bits.

This information is required in the AllStubs to be used for the new VMSMERouter dual FPGA module, and it was decided this was the easiest way to incorporate the information without making changes to the previous processing modules / causing the DTC/Input/AllStubs to be of different format. More information regarding this is available here.

PR also includes changes to written memory text files required for current master firmware-hls (writing offset r values for DiskPS stubs and negDisk bit for all disk stubs).

aryd committed:
First pass at removing non-combined modules

aryd committed:
Removed unused iMath code

aryd committed:
First code for a ProjectionCalculator module

Commit:
Avoid stale pointers on subsequent events

Commit:
Cleaner MP pointer checks

Full agreement with HLS (100 events L1PHIC - D5PHIC)

aryd committed:
Set of changes to make a configuration that uses the Projection Calculator module

aryd committed:
Fix mistake in code merging

tomalin committed:
Andrews KF crash fix (#263)

tomalin committed:
Add z0 resolution to performance printout (#264)

* Add z0 resolution to performance printout

* code format

tomalin committed:
DR: fix displaced track bug & disable binning (#266)

* Disable binning in DR

* bug fix

* add comment

* code format

* tweak comment

* fix previous erroneous commit

tomalin committed:
DUMMY COMMIT BEFORE PR TO CENTRAL CMSSW

tomalin committed:
Run combined modules by default (#265)

* Make combined modules default

* tweak

* Improve USEHYBRID ifdef range

* Fix compiler error for pure Tracklet algo

* Move fitpattern.txt refs, so only used for pure Tracklet algo

* code format

aehart committed:
Numerical stability fix (#269)

* Made calculations more numerically stable.

* Explicitly restrict to domain of asin/acos.

* Added comments.

* Code format.

* Fixed typos in comments.

* Only do calculations when needed.

* Added const where applicable.

tomalin committed:
Fix inventStubs bug in duplicate removal (#271)

* Fix inventStubs bug in duplicate removal

When using L2L3D1 seeds, it is only L2L3 that are used to determined the r-z helix params, not L2D1 as the code currently assumes.

* Update PurgeDuplicate.cc

* add comment

* formatted

aryd committed:
Changes needed for the VMSME Router module

mlarson02 committed:
Fully implementing VMSMERouter, VMRouterCM no longer producing VMMEStubs

mlarson02 committed:
Manually incorporating DTC stub, TT stub changes to CMSSW 14 dev branch

mlarson02 committed:
Manually incorporating DTC stub, TT stub changes to CMSSW 14 dev branch

mlarson02 committed:
Readding TCBase, VMSMER compatible w/ updated stub format

mlarson02 committed:
Merge branch 'ProjectionCalculator_without_IMATH_and_noncombined' into StubsCMSSW14

aryd commented on L1Trigger/TrackFindingTMTT/src/KFbase.cc:
```diff
@@ -551,7 +551,7 @@ namespace tmtt {
     } else {
       // Protection against rare maths instability.
-      const TMatrixD unitMatrix(TMatrixD::kUnit, TMatrixD(nHelixPar_, nHelixPar_));
+      const TMatrixD unitMatrix(TMatrixD::kUnit, TMatrixD(2, 2));
```
Is this related to these changes?

aryd commented on L1Trigger/TrackFindingTracklet/README.md:
```diff
@@ -1,4 +1,4 @@
-To run the L1 tracking & create a TTree of tracking performance:
+To run the L1 tracking & create a TTree of tracking performance: 
```
Remove extra space

aryd commented on L1Trigger/TrackFindingTracklet/interface/Settings.h:
```diff
@@ -344,7 +345,7 @@ namespace trklet {
     double kz() const { return 2.0 * zlength_ / (1 << nzbitsstub_[0]); }
     double kz(unsigned int layerdisk) const { return 2.0 * zlength_ / (1 << nzbitsstub_[layerdisk]); }
-    double kr() const { return rmaxdisk_ / (1 << nrbitsstub_[N_LAYER]); }
+    double kr() const { return rmaxdisk_ / (1 << (nrbitsstub_[N_LAYER] + 1)); }
```
Can you add a comment to explain?

aryd commented on L1Trigger/TrackFindingTracklet/interface/Settings.h:
```diff
@@ -1051,18 +1053,23 @@ namespace trklet {
     //Rinv bins were optimised to ensure a similar number of tracks in each bin prior to DR
     //Rinv bin edges for 6 bins.
-    std::vector<double> rinvBins_{-rinvcut(), -0.004968, -0.003828, 0, 0.003828, 0.004968, rinvcut()};
+    //std::vector<double> rinvBins_{-rinvcut(), -0.004968, -0.003828, 0, 0.003828, 0.004968, rinvcut()};
```
Are these changes related to this updated?

aryd commented on L1Trigger/TrackFindingTracklet/interface/Tracklet.h:
```diff
@@ -221,6 +221,8 @@ namespace trklet {
 
     unsigned int PSseed() const { return ((layer() == 1) || (layer() == 2) || (disk() != 0)) ? 1 : 0; }
 
+    // Seed type:
```
Why is this added?

aryd commented on L1Trigger/TrackFindingTracklet/src/Stub.cc:
```diff
@@ -160,7 +164,7 @@ double Stub::rapprox() const {
       return settings_.rDSSouter(r_.value());
   }
-  return r_.value() * settings_.kr();
+  return (r_.value() + (1 << 8)) * settings_.kr();  // Incorporating r offset for disk PS stubs
```
Can you do something like "int r_offset_bits=8" and use this variable?

aryd commented on L1Trigger/TrackFindingTracklet/src/TrackletProcessor.cc:
```diff
@@ -403,9 +403,9 @@ void TrackletProcessor::execute(unsigned int iSector, double phimin, double phim
           indexz = ((1 << nbitszfinebintable_) - 1) - indexz;
         }
-        indexr = stub->r().value() >> (stub->r().nbits() - nbitsrfinebintable_);
+        indexr = stub->rvalue() >> (stub->r().nbits() + 1 - nbitsrfinebintable_);
```
Can you add comment about offset of +1

aryd reviewed (commented):
I added a few questions and comments on the updates.

mlarson02 committed:
Addressing comments, fixing code format

mlarson02 commented:
> I added a few questions and comments on the updates.

I've addressed the comments relating to my changes, the others are regarding code included in CMSSW 14 dev branch commits incorporated in my branch, but not yet merged into the ProjectionCalculator_without_IMATH_and_noncombined branch.

mlarson02 committed:
Removing unneeded diskpswrittenr variable

mlarson02 committed:
Removing unneeded + 1 to rbits in TP

aryd committed:
Updates to make sure we don't have missing projections
